### DISCUSSION TITLE
Use cases for OpenAPI

### DISCUSSION THREAD

**Discussion by darrelmiller** [OAI/sig-moonwalk#130] [Ideas]
As part of the moonwalk TDC meeting of 2024-06-04 there was an acknowledgement that it would be valuable to capture the set of use cases that OpenAPI descriptions need to satisfy.

OpenAPI Tooling Developer Use Cases
- Request matching 
- Operation Description
- Security definition
- Request Construction
- Response Handling
- Compliance
  -  Linting
  - OpenAPI Validation
- Authoring

Application Developer use cases (API consumption and production)
- Payload validation
- Routing
- Request validation
- Response validation
- Metrics collections
- Quota enforcement
- Test generation 
- Compliance
  -  Linting

- Documentation generation
- API Client code generation

This is brainstorm of ideas, do not consider it final or even partially correct. We will continue iterating on this next week.

**Comment by garymazz** [comment]
Additional use cases under Application Developer cluster

Authoring:

- Support the use of multiple API definition files for separating technical concerns and parallel definition development
- Allow inclusion/import of one or more definition files to create a API definition 
- Allow included/imported definition files to inclusion/import of one or more definition files
- Support local file access to included/imported definition files
- Support web access to included/imported definition files included/imported definition files
- Use objects and properties defined in included/imported definition files to create an API definition
- Detect and notify (warning/error) developer of conflicts between identically named objects with dissimilar definitions defined in included definition files
- Detect and notify (warning/error) developer of conflicts between identically named properties with dissimilar definitions defined in included definition files
- Support inclusion of external schemes to create an API definition. eg geojson
- Support author augmentation of previously defined and included/imported objects and properties
- Support author override of previously defined and included/imported "required" objects properties 
- Detect and notify (warning/error) author of augmented included/imported objects and properties
- Detect and notify (warning/error) author of overridden "required" objects properties 
- Support author selected reporting of detected conflicts between identically named properties with dissimilar definitions as ignored/warning/error
- Support author selected reporting of detected conflicts between identically named objects with dissimilar definitions as ignored/warning/error
- Support author selected reporting of detected author augmented objects and properties as ignored/warning/error
- Support author selected reporting of detected author overridden "required" objects properties as ignored/warning/error
- Report inaccessible included/imported definition files

**Comment by garymazz** [comment]
Abstract definitions: rough draft
**Modularity:** 
The ability to separate the API definition into elements or components which may also be separates into other elements or components. Modularity may include API definition files, paths, schemas, and all other named object and properties

**Extensibility:**
The ability to augment and extend previously defined objects and properties via overriding existing objects and properties definitions. 

**Composability:**
The ability to define/construct objects and properties from previously defined objects and properties. Constructed objects and properties  can reference  previously defined objects and properties by unique name or other unique identifier.  Objects and properties may define rules permitting and limiting the types or names/identifiers of previously defined objects and properties used with constructed objects' and properties' definition.
